libsignal-protocol/src/stores/session_store.rs

```diff
@@ -0,0 +1,204 @@
+            }
+
+            1
+        },
+        Ok(None) => 0,
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn get_sub_device_sessions_func(
+    sessions: *mut *mut sys::signal_int_list,
+    name: *const c_char,
+    name_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!sessions.is_null());
+    assert!(!name.is_null());
+    assert!(!user_data.is_null());
+
+    let state = &*(user_data as *const State);
+    let name = std::slice::from_raw_parts(name as *const _, name_len);
+
+    match state.0.get_sub_device_sessions(name) {
+        Ok(got) => unimplemented!(),
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn store_session_func(
+    address: *const sys::signal_protocol_address,
+    record: *mut u8,
+    record_len: usize,
+    user_record: *mut u8,
+    user_record_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!address.is_null());
+    assert!(!record.is_null());
+    assert!(!user_data.is_null());
+
+    let state = &*(user_data as *const State);
+    let addr = Address::from_ptr(address);
+    let record = std::slice::from_raw_parts(record, record_len);
+    let user_record = if user_record.is_null() {
+        None
+    } else {
+        Some(std::slice::from_raw_parts(user_record, user_record_len))
+    };
+
+    let session = SerializedSession {
+        session: Buffer::from(record),
+        extra_data: user_record.map(Buffer::from),
+    };
+
+    match state.0.store_session(addr, session) {
+        Ok(_) => sys::SG_SUCCESS as _,
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn contains_session_func(
+    address: *const sys::signal_protocol_address,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!address.is_null());
+    assert!(!user_data.is_null());
+
+    let state = &*(user_data as *const State);
+    let addr = Address::from_ptr(address);
+
+    match state.0.contains_session(addr) {
+        Ok(true) => 1,
+        Ok(false) => 0,
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn delete_session_func(
+    address: *const sys::signal_protocol_address,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!address.is_null());
+    assert!(!user_data.is_null());
+
+    let state = &*(user_data as *const State);
+    let addr = Address::from_ptr(address);
+
+    match state.0.delete_session(addr) {
+        Ok(_) => sys::SG_SUCCESS as _,
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn delete_all_sessions_func(
+    name: *const c_char,
+    name_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!name.is_null());
+    assert!(!user_data.is_null());
+
+    let state = &*(user_data as *const State);
+    let name = std::slice::from_raw_parts(name as *const _, name_len);
+
+    match state.0.delete_all_sessions(name) {
+        Ok(_) => sys::SG_SUCCESS as _,
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn destroy_func(user_data: *mut c_void) {
```

Comment:
do we have to check if `user_data` isn't null here ?

Reply:
I think it'd be good practice to. I inserted the `assert!(!user_data.is_null())` because we later dereference it.